### ISSUE TITLE
YOLO Model Fails to Detect All Classes in Multi-Class Inference Image

### ISSUE THREAD

**Issue by HGSChandeepa** [ultralytics/ultralytics#21609]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I trained a YOLO model on a **dataset of single-class images** (e.g., individual snack packets) and attempted **inference on an image containing multiple classes** (e.g., Lay’s, Sweet Potato Chips, Cheese Curls, etc.). 

The model successfully detects some classes but fails to detect all classes present in the multi-class image. Despite adjusting training configurations with augmentations (mosaic, mixup, erasing) and increasing epochs, the issue persists. This suggests a potential limitation in generalizing from single-class training data to multi-class inference scenarios.

`target_model.train(
    data="/content/yolo_dataset/data.yaml",
    epochs=50,
    patience=20,
    batch=8,
    imgsz=640,
    save=True,
    device=0,
    project="YOLOv8_Runs",
    name="exp_with_enhanced_augment",
    optimizer='AdamW',
    lr0=0.001,
    lrf=0.01,
    hsv_h=0.02,
    hsv_s=0.5,
    hsv_v=0.4,
    degrees=15.0,
    translate=0.2,
    scale=0.4,
    fliplr=0.5,
    mosaic=1.0,
    mixup=0.3,
    erasing=0.5,
    shear=0.1
)`

Expected Behavior

The model should detect all classes (Lay’s, Sweet Potato Chip, Cheese Curls, etc.) present in the multi-class inference image with accurate bounding boxes and confidence scores.

Actual Behavior

The model detects only a subset of the classes (e.g., Lay’s and Sweet Potato Chip) and fails to detect others (e.g., Cheese Curls, Star Ring), even with a lowered confidence threshold (0.3).

Please provide guidance on improving the model’s ability to detect all classes in multi-class inference images or suggest debugging steps to identify the root cause.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @HGSChandeepa, thank you for reaching out and sharing your experience with Ultralytics 🚀! This is an automated response to help you get the most effective support possible. An Ultralytics engineer will review your issue and assist you soon.

We recommend checking out the [Docs](https://docs.ultralytics.com/) for detailed usage guides. For training-specific questions, please ensure you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Since this issue may relate to model or dataset configuration, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) if possible. This helps us debug more efficiently. Include:
- A small sample of your dataset
- Your data.yaml and any custom scripts/configs
- Training/inference logs or screenshots showing the problem

Join the Ultralytics community for support and discussions:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth topics: [Discourse](https://community.ultralytics.com/)
- Community threads: [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Please ensure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO models can be run in any of these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for helping us improve Ultralytics!

**Comment by Y-T-G** [comment]
How many images do you have? Can you post your `labels.jpg`?

**Comment by HasnatHridoy** [comment]
Oh man, I am having the same issue.

**Comment by glenn-jocher** [comment]
@HasnatHridoy thanks for confirming—please open a new issue referencing this thread (it’s marked Stale) so we can track your case; include diagnostics, a minimal repro, and text logs. To move this forward: run system checks, share `model.val(..., plots=True)` outputs (per-class PR/confusion), and note that `erasing` is classification-only so it won’t affect detect training; adding some multi-object training images and testing a larger `imgsz` at inference typically resolves the “single-object trained → multi-class scene” gap. See [Troubleshooting Common YOLO Issues](https://docs.ultralytics.com/guides/yolo-common-issues/), validation args in [Model Validation with Ultralytics YOLO](https://docs.ultralytics.com/modes/val/), and the note on Random Erasing in the [Data Augmentation guide](https://docs.ultralytics.com/guides/yolo-data-augmentation/).

```bash
yolo checks
```

```python
from ultralytics import YOLO
model = YOLO("best.pt")
model.val(data="path/to/data.yaml", plots=True)
```

**Comment by HGSChandeepa** [comment]
@HasnatHridoy @glenn-jocher 

In our case, the major issue was with the dataset.
We trained the model using images containing **only a single class**, but during inference we used images with **multiple classes**. Since the model was trained to detect only a single class, this mismatch caused performance issues.

### My Solution

I wrote a script to generate **synthetic data** using different backgrounds and placing multiple images on them.

* These scripts can be used for **both detection and segmentation tasks**.
* They also return the data in the **exact format required to train YOLO models** for detection and segmentation.

🔗 Script & setup instructions:
[[GitHub – Synthetic Data Generation](https://github.com/HGSChandeepa/Synthetic-data-generation)](https://github.com/HGSChandeepa/Synthetic-data-generation)

### Additional Tips for Improving Accuracy

* Apply **data augmentation** to diversify the dataset.
* Use **SAHI (Sliced Inference)** for handling large input images and improving detection.

📚 References:

* [[YOLO Data Augmentation Guide](https://docs.ultralytics.com/guides/yolo-data-augmentation/)](https://docs.ultralytics.com/guides/yolo-data-augmentation/)
* [[SAHI Tiled Inference Guide](https://docs.ultralytics.com/guides/sahi-tiled-inference/)](https://docs.ultralytics.com/guides/sahi-tiled-inference/)

**Comment by HasnatHridoy** [comment]
@HGSChandeepa  Thank you.... I will definitely try this out. In the meantime, I found the 'RF-detr' models somehow achieve very good results for single-object trained → multi-class scene detection. All you need just lower the threshold.
